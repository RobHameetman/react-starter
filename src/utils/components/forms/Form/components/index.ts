export * from './CloseButton';
export * from './EditButton';
export * from './FormButtons';
export * from './FormField';
export * from './SubmitButton';
