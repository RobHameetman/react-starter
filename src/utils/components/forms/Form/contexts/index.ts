export * from './FormContext';
export * from './FormProvider';
