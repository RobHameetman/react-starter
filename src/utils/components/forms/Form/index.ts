export * from './components';
export * from './contexts';
export * from './enums';
export * from './functions';
export * from './hooks';
export * from './types';

export * from './Form';
