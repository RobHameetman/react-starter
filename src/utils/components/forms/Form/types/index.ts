export * from './CachedForm';
export * from './FormField';
export * from './FormFields';
