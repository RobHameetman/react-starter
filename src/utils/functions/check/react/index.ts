export * from './isChangeEvent';
export * from './isDragEvent';
export * from './isEventHandler';
export * from './isFocusEvent';
export * from './isFocusEventHandler';
export * from './isKeyboardEvent';
export * from './isMouseEvent';
export * from './isMutableRefObject';
export * from './isPointerEvent';
export * from './isReactElement';
export * from './isScrollEvent';
export * from './isSyntheticEvent';
export * from './isUIEvent';
export * from './isUIEventHandler';
