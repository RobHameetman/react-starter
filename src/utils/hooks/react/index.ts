export * from './useControlProp';
export * from './useEnterEvents';
export * from './useEscapeEvents';
export * from './useFocusEvents';
export * from './usePressEvents';
export * from './usePropsWithChildren';
export * from './useSemanticAsProp';
export * from './useSpaceEvents';
export * from './useTabEvents';
