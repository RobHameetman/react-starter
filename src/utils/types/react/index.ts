export * from './CC';
export * from './CompoundComponent';
export * from './HandlerPropName';
export * from './HandlerPropsOf';
export * from './PropsWithoutHandlers';
