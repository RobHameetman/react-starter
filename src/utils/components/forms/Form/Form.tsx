import { FormEventHandler, MouseEvent, useCallback } from 'react';
import { noop } from '@app/utils/functions/misc/noop';
import { CC } from '@app/utils/types/react/CC';
import type { Stylable } from '@app/utils/types/props/Stylable';
import {
	FormButtons,
	CloseButton,
	EditButton,
	FormField,
	SubmitButton,
} from './components';
import { FormProvider } from './contexts';
import { FormModes } from './enums';

/**
 * Compositional prop types for the {@link FormButtons} component.
 */
type ComposedProps = Stylable;

/**
 * Prop types for the {@link Form} component.
 */
export interface FormProps extends ComposedProps {
	/**
	 * [Optional] Override the default form mode.
	 * @defaultValue - `FormModes.Write`
	 */
	readonly defaultMode?: FormModes;

	/**
	 * [Optional] Set to `true` to prevent the form state from persisting in LocalStorage.
	 * @defaultValue - `false`
	 */
	readonly disableCache?: boolean;

	/**
	 * [Optional] Set to `false` to make the form "write-only".
	 * @defaultValue - `true`
	 */
	readonly editable?: boolean;

	/**
	 * A name for the form used for caching.
	 */
	readonly name: string;

	/**
	 * [Optional] A name for the form used for caching.
	 * @defaultValue - A no-op function.
	 */
	readonly onSubmit?: FormEventHandler;
}

export interface FormComponents {
	Buttons: typeof FormButtons;
	Close: typeof CloseButton;
	Edit: typeof EditButton;
	Field: typeof FormField;
	Submit: typeof SubmitButton;
}

/**
 * A short description of the component here.
 */
export const Form: CC<FormComponents, FormProps> = ({
	className = '',
	children,
	defaultMode = FormModes.Write,
	disableCache = false,
	editable = true,
	name,
	onSubmit = noop,
}) => {
	const cssOverrides = className ? ` ${className}` : '';

	const _handleSubmit = useCallback((e: MouseEvent<HTMLFormElement>) => {
		e.preventDefault();

		onSubmit(e);
	}, []);

	return (
		<FormProvider
			name={name}
			disableCache={disableCache}
			editable={editable}
			mode={editable ? defaultMode : FormModes.Write}
		>
			<form className={cssOverrides} onSubmit={_handleSubmit}>
				{children}
			</form>
		</FormProvider>
	);
};

Form.Buttons = FormButtons;
Form.Close = CloseButton;
Form.Edit = EditButton;
Form.Field = FormField;
Form.Submit = SubmitButton;
