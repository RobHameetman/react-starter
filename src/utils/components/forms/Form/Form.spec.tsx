import { render } from '@testing-library/react';
import { Form } from './Form';

describe('<Form />', () => {
	it('should render correctly', () => {
		expect(() => render(<Form name="test" />)).not.toThrowError();
	});

	it('should not regress', () => {
		expect(() => render(<Form name="test" />)).toMatchSnapshot();
	});
});
