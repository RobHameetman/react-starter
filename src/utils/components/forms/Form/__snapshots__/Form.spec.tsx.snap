// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should not regress 1`] = `undefined`;
